pkg/apiserver/webhooks.go

```diff
@@ -226,29 +238,130 @@ func createValidatingWebhook(ctx context.Context, webhookNs string, caCert []byt
+func loadCertificate(certPath, keyPath string) (tls.Certificate, error) {
+	// get info about cert manager certificate secret mounted as a volume on the porch server pod
+	certInfo, err := os.Stat(certPath)
 	if err != nil {
-		return err
+		return tls.Certificate{}, err
+	}
+	// if the last time this secret was modified was after the last time we loaded its cert files
+	// we lock access to the mount path and load the keypair from the files in our tls then release the lock
+	// set this new loaded cert as our current cert and note the modtime for next reload
+	if certInfo.ModTime().After(certModTime) {
+		mu.Lock()
+		defer mu.Unlock()
+		newCert, err := tls.LoadX509KeyPair(certPath, keyPath)
+		if err != nil {
+			return tls.Certificate{}, err
+		}
+		cert = newCert
+		certModTime = certInfo.ModTime()
 	}
-	klog.Infoln("Starting webhook server")
-	http.HandleFunc(serverEndpoint, validateDeletion)
-	server := http.Server{
-		Addr: fmt.Sprintf(":%d", webhookServicePort),
-		TLSConfig: &tls.Config{
-			Certificates: []tls.Certificate{cert},
-		},
+	return cert, nil
+}
+
+// watch for changes on the mount path of the secret as volume
+func watchCertificates(directory, certFile, keyFile string) {
+	//set up a watcher for the cert
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.Fatalf("failed to start certificate watcher: %v", err)
 	}
+	defer watcher.Close()
+
+	// if the watcher notices any changes on the mount dir of the secret such as creations or writes to the files in this dir
+	// attempt to load tls cert using the new cert and key files provided and log output
+	done := make(chan bool)
 	go func() {
-		err = server.ListenAndServeTLS("", "")
-		if err != nil {
-			klog.Errorf("could not start server: %v", err)
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if event.Op&fsnotify.Create == fsnotify.Create || event.Op&fsnotify.Write == fsnotify.Write {
+					_, err := loadCertificate(certFile, keyFile)
+					if err != nil {
+						klog.Errorf("Failed to load updated certificate: %v", err)
+					} else {
+						klog.Info("Certificate reloaded successfully")
+					}
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				klog.Errorf("Watcher error: %v", err)
+			}
 		}
 	}()
-	return err
+	// start the watcher with the dir to watch
+	err = watcher.Add(directory)
+	if err != nil {
+		klog.Fatalf("Error in running watcher: %v", err)
+	}
+
+	<-done
+}
+// return current cert
+func getCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	mu.Lock()
+	defer mu.Unlock()
+	return &cert, nil
+}
+
+func runWebhookServer(certStorageDir string, useCertManWebhook bool) error {
+	certFile := filepath.Join(certStorageDir, "tls.crt")
+	keyFile := filepath.Join(certStorageDir, "tls.key")
+	// load the cert for the first time
+
+	if useCertManWebhook {
+		_, err := loadCertificate(certFile, keyFile)
```

Comment:
This gets called within the watchCerts function again. Why do we need to call it here too?
Also, should we exit below if an error is thrown?

Reply:
- this line is the first time the certificate gets loaded if we are using cert manager operation.
- i gets called again in the watchCerts function because in the case where a new certificate is detected aka the secret from cert manager has been updated with new contents, we want to reload the certificate with the new tls.crt & tls.crt contents.
- as for the exit if error i believe so, since if we cannot even load the certificate on the attempt there is not much point in attempting to start the server without any tls? or am i misunderstanding are you suggesting a retry for 2/3 times before denial?